torch/jit/_recursive.py

```diff
@@ -321,7 +334,7 @@ def get_module_concrete_type(nn_module, share_types=True):
 
     return concrete_type
 
-def create_script_module(nn_module, stubs_fn, share_types=True):
+def create_script_module(nn_module, stubs_fn, share_types=True, reuse_stubs_fn=False):
```

Comment:
As far as I can tell, `reuse_stubs_fn` is always `True` right? Because either:
1. You are tracing, and you want to keep tracing the forwards of the other modules
2. You are scripting, and you are using the default `stubs_fn`.

Either way, the `stubs_fn` is the same within a given recursive run, right? If that's the case, can we just remove this option and ensure that we always re-use the same stubs fn?

Reply:
While this is almost true and it was the methodology that I was trying, but unfortunately it fails, because we have one exception: the Legacy `ScriptModule` compilation, if you look at here https://github.com/pytorch/pytorch/blob/7816d53798efcd1dc68a83a2964ee6d4ef4181ca/torch/jit/_script.py#L203 we are using the `make_stubs` rule to compile the legacy ScriptModule as entry point, but inside the recursive run, we are using the default `infer_methods_to_compile` rule for submodule compilation. I think this is because `cls._methods.items()` is not available in ScriptModule's submodule bc that might be a plain nn.Module and don't have `cls._methods` crafted upon construction. 

In fact, if I always use the same `stubs_fn`, it will crash on some tests and complain something like this:

```
  File "/home/wanchaol/pytorch/torch/jit/_recursive.py", line 364, in create_script_module_impl
    method_stubs = stubs_fn(nn_module)
  File "/home/wanchaol/pytorch/torch/jit/_script.py", line 202, in make_stubs
    return [v for k, v in sorted(cls._methods.items())]
AttributeError: type object 'Over' has no attribute '_methods'
```


I think we should either: 1. use the approach I proposed here 2. disable recursive compilation in legacy ScriptModule API, only allow user to construct their submodule with all modules inheriting from `torch.jit.ScriptModule` (or say, `torch.jit.ScriptModule` should only contain submodule with inheriting from `torch.jit.ScriptModule`.)

I would prefer the first option bc it does not BC break the behavior that we already exposed to the user.

Reply:
In general, I would prefer to add more to our legacy path shims, rather than change the current behavior just because the legacy path has some constraints. That way we can keep the current behavior clean and add the complexity in the legacy path. 

In this case, that would mean: change the custom `make_stubs` used in the the legacy ScriptModule compilation to fork its behavior, so it would look like:
```
def make_stubs:
     if I am compiled the legacy way:
           look in _methods to make stubs
     else:
           use the regular `infer_methods_to_compile`
```